File: kernels/triton/fused_attn.py
# Triton baseline for fused QKV attention.
#
# This is the reference implementation used for:
# 1. Correctness validation against kernels/cuda_oxide/src/fused_attn.rs
# 2. Performance comparison in tests/benchmark/
#
# PyTorch reference: fused_attn_torch() below
# Rust counterpart:  kernels/cuda_oxide/src/fused_attn.rs

import math

import torch
import triton
import triton.language as tl


@triton.jit
def _fused_attn_kernel(
    qkv_ptr,
    out_ptr,
    stride_b,
    stride_s,
    seq_len,
    num_heads,
    head_dim,
    scale,
    BLOCK_M: tl.constexpr,
    BLOCK_N: tl.constexpr,
):
    # TODO: implement Triton fused attention kernel (FlashAttention-style tiling)
    # 1. pid_b = program_id(0), pid_h = program_id(1), pid_m = program_id(2)
    # 2. Load Q tile from qkv_ptr (first head_dim cols)
    # 3. Loop over K/V tiles, compute softmax(QK^T * scale) online (numerically stable)
    # 4. Accumulate weighted V
    # 5. Store output tile
    pass


def fused_attn_triton(
    qkv: torch.Tensor,
    num_heads: int,
    head_dim: int,
) -> torch.Tensor:
    """Launch the Triton fused attention kernel.

    Args:
        qkv:       Packed tensor of shape [batch, seq_len, 3 * num_heads * head_dim], fp32, CUDA
        num_heads: Number of attention heads
        head_dim:  Dimension per head (must be power of 2, <= 128 for sm_80)

    Returns:
        Attention output of shape [batch, seq_len, num_heads * head_dim]
    """
    # TODO: validate shapes, compute grid, launch _fused_attn_kernel
    raise NotImplementedError("fused_attn_triton not yet implemented")


def fused_attn_torch(
    qkv: torch.Tensor,
    num_heads: int,
    head_dim: int,
) -> torch.Tensor:
    """PyTorch reference implementation (used for correctness checks)."""
    batch, seq_len, _ = qkv.shape
    scale = 1.0 / math.sqrt(head_dim)
    q, k, v = qkv.split(num_heads * head_dim, dim=-1)
    q = q.view(batch, seq_len, num_heads, head_dim).transpose(1, 2)
    k = k.view(batch, seq_len, num_heads, head_dim).transpose(1, 2)
    v = v.view(batch, seq_len, num_heads, head_dim).transpose(1, 2)
    attn = torch.softmax(torch.matmul(q, k.transpose(-2, -1)) * scale, dim=-1)
    out = torch.matmul(attn, v)
    return out.transpose(1, 2).contiguous().view(batch, seq_len, num_heads * head_dim)

File: kernels/cuda_oxide/src/fused_attn.rs
//! Fused attention kernel (sm_80, A100).
//!
//! Implements fused QKV projection + scaled dot-product attention in a single kernel,
//! avoiding intermediate materialization of Q, K, V tensors.
//!
//! References:
//! - FlashAttention: <https://arxiv.org/abs/2205.14135>
//! - Triton baseline: `kernels/triton/fused_attn.py`

use thiserror::Error;

#[derive(Debug, Error)]
pub enum FusedAttnError {
    #[error("sequence length {0} exceeds maximum supported {1}")]
    SeqLenExceeded(usize, usize),
    #[error("head_dim must be a power of 2, got {0}")]
    InvalidHeadDim(usize),
    #[error("CUDA kernel launch failed: {0}")]
    CudaLaunch(String),
}

/// Fused QKV attention kernel wrapper (sm_80).
pub struct FusedAttn {
    num_heads: usize,
    head_dim: usize,
    scale: f32,
}

impl FusedAttn {
    /// Create a new FusedAttn kernel.
    ///
    /// * `num_heads` — number of attention heads
    /// * `head_dim`  — dimension per head (must be power of 2, ≤ 128 for sm_80)
    pub fn new(num_heads: usize, head_dim: usize) -> Self {
        let scale = 1.0 / (head_dim as f32).sqrt();
        Self {
            num_heads,
            head_dim,
            scale,
        }
    }

    /// Forward pass: compute attention output from packed QKV input.
    ///
    /// # Arguments
    /// * `qkv` — packed tensor of shape `[batch, seq_len, 3 * num_heads * head_dim]`
    ///
    /// # Errors
    ///
    /// Returns [`FusedAttnError::SeqLenExceeded`] if seq_len > 2048 (sm_80 shared memory limit).
    /// Returns [`FusedAttnError::InvalidHeadDim`] if head_dim is not a power of 2.
    pub fn forward(&self, _qkv: &[f32]) -> Result<Vec<f32>, FusedAttnError> {
        // TODO: implement fused attention cuda-oxide kernel
        // 1. Validate seq_len <= 2048 (tiling constraint for sm_80 SMEM budget: 164 KB)
        // 2. Validate head_dim is power of 2
        // 3. Allocate device buffers: qkv_dev, output_dev
        // 4. Copy host → device
        // 5. Launch fused_attn_kernel<<<grid, block, smem>>>(qkv, output, scale, seq_len, head_dim)
        // 6. Copy output device → host
        let _ = (self.num_heads, self.head_dim, self.scale);
        todo!("fused_attn forward pass not yet implemented")
    }
}
